docs/adr/002-watch-synchronization.md

```diff
@@ -0,0 +1,160 @@
+# 002 Watch Synchronization
+
+Status:  Accepted
+
+Date: 2025-04-04
+
+Tags: watch, synchronization, architecture
+
+## Context
+
+**Problem Statement:** We need to ensure that the watch synchronization mechanism in our system is efficient, effective and reliable.
+
+**Previous Implementation (Brief):** The previous implementation handling Kubernetes watch events inherited from Kine was not designed for a single synchronization point of all watch streams. The system used a broadcaster to manage watch stream subscriptions and forward events from a poll loop. It follows the fan-out and fan-in go pattern for distributing the watch events to the watchers which send the watched events to the API server. After the WatchList feature in Kubernetes upstream required an additional watch progress notification a patch was made on top of the event implementation waiting on events to achieve this synchronization point. An issue with a missing event caused us to revisit the implementation.
```

Comment:
```suggestion
**Previous Implementation (Brief):** The previous implementation handling Kubernetes watch events inherited from Kine was not designed for a single synchronization point of all watch streams. The system used a broadcaster to manage watch stream subscriptions and forward events from a poll loop. It follows the fan-out and fan-in Go pattern for distributing the watch events to the watchers which in turn send the watched events to the API server. After the WatchList feature in Kubernetes upstream required an additional watch progress notification, a patch was made on top of the event implementation waiting on events to achieve this synchronization point. An issue with a missing event caused us to revisit the implementation.
```

Reply:
I am a bit confused by this part "a patch was made on top of the event implementation waiting on events to achieve this synchronization point" Could you clarify?